src/unaudited/EcdsaEpochRegistry_Permissionless.sol

```diff
@@ -0,0 +1,250 @@
+        uint256 _maxRegisteredOperators,
+        uint256 _defaultWeightRequirement,
+        uint256 _retargetingFactorWei
+    )
+        EcdsaEpochRegistry_Modular(_serviceManager, _delegationManager, _epochLengthSeconds, _epochZeroStart)
+    {
+        require(_targetOperatorSetSize != 0, "no.");
+        require(_maxRegisteredOperators > _targetOperatorSetSize, "max must be above target");
+        minimumWeightRequirement = _minimumWeightRequirementInitValue;
+        emit MinimumWeightChanged(_minimumWeightRequirementInitValue);
+        targetOperatorSetSize = _targetOperatorSetSize;
+        maxRegisteredOperators = _maxRegisteredOperators;
+        defaultWeightRequirement = _defaultWeightRequirement;
+        retargetingFactorWei = _retargetingFactorWei;
+    }
+
+    /*******************************************************************************
+                    EXTERNAL FUNCTIONS - unpermissioned
+    *******************************************************************************/
+    /**
+     * @notice Registers msg.sender as an operator.
+     * @param operatorSignature is the signature of the operator used by the AVS to register the operator in the delegation manager
+     */
+    function registerOperator(
+        ISignatureUtils.SignatureWithSaltAndExpiry memory operatorSignature
+    ) public virtual override {
+        address operator = msg.sender;
+        _registerOperator(operator, operatorSignature);
+    }
+
+    // @notice Deregisters the msg.sender.
+    function deregisterOperator() public virtual override {
+        address operator = msg.sender;
+        _deregisterOperator(operator);
+    }
+
+    // TODO: documentation
+    // evaluates the operator set for the current epoch
+    function evaluateOperatorSet(
+        address[] memory /*operatorsToEvaluate*/,
+        uint256 /*minimumWeightRequirement*/
+    ) public virtual override {
+        uint256 _currentEpoch = currentEpoch();
+        require(operatorsForEpoch[_currentEpoch].length == 0, "operator set already calculated for current epoch");
+
+        uint256 _numberRegisteredOperators = numberRegisteredOperators();
+        uint256 _realizedMinimumWeight = type(uint256).max;
+
+        for (uint256 i = 0; i < _numberRegisteredOperators; ++i) {
+            address operator = registeredOperators[i];
+            uint256 operatorWeight = weightOfOperator(operator);
+            // add any registered operator that meets the weight requirement to the current epoch's operator set
+            if (operatorWeight >= minimumWeightRequirement) {
+                operatorsForEpoch[_currentEpoch].push(operator);
+                _operatorStakeHistory[operator][_currentEpoch] = operatorWeight;
+                // TODO: add event
+                if (operatorWeight < _realizedMinimumWeight) {
+                    _realizedMinimumWeight = operatorWeight;
+                }
+            // otherwise, evict the operator
+            } else {
```

Comment:
It seems unnecessary to kick the operators here.

Then, they may be considered in a new epoch without any additional action, either by retargeting deregistration of other operators, or new delegations. They might meet the minimum requirements.  

This could complicate payments?

Reply:
hmm, good point that if the target goes down, you might not want to kick people before then.
I believe my thought process was "we are checking if they meet the current requirement, then we will retarget".
will give the optimal design more thought though.